File: archive/extract_schema.mjs
import sqlite3 from "sqlite3";
/**
 * Utility script for extracting a readable schema summary from SQLite.
 */

/**
 * Promise-based wrapper around sqlite3 `db.all`.
 * @param {sqlite3.Database} db
 * @param {string} sql
 * @returns {Promise<any[]>}
 */
function all(db, sql) {
  return new Promise((resolve, reject) => {
    db.all(sql, (err, rows) => {
      if (err) {
        reject(err);
        return;
      }
      resolve(rows);
    });
  });
}

/**
 * Close sqlite database connection.
 * @param {sqlite3.Database} db
 * @returns {Promise<void>}
 */
function close(db) {
  return new Promise((resolve, reject) => {
    db.close((err) => {
      if (err) {
        reject(err);
        return;
      }
      resolve();
    });
  });
}

/**
 * Open a SQLite database from path.
 * @param {string} dbPath
 * @returns {Promise<sqlite3.Database>}
 */
function openDatabase(dbPath) {
  return new Promise((resolve, reject) => {
    const db = new sqlite3.Database(dbPath, (err) => {
      if (err) {
        reject(err);
        return;
      }
      resolve(db);
    });
  });
}

/**
 * Return a formatted schema summary for all tables/columns in the database.
 * @param {string} dbPath
 * @returns {Promise<string>}
 */
export async function getDatabaseSchema(dbPath) {
  const db = await openDatabase(dbPath);

  try {
    const tables = await all(db, "SELECT name FROM sqlite_master WHERE type='table';");
    const schemaLines = [];

    for (const table of tables) {
      const tableName = table.name;
      schemaLines.push(`Table name: ${tableName}`);
      schemaLines.push("Columns:");

      const safeTableName = tableName.replace(/'/g, "''");
      const columns = await all(db, `PRAGMA table_info('${safeTableName}');`);

      for (const col of columns) {
        const colType = col.type || "TEXT";
        schemaLines.push(`  - ${col.name} (${colType})`);
      }

      schemaLines.push("");
    }

    return schemaLines.join("\n");
  } finally {
    await close(db);
  }
}

// --- Example usage when invoked directly ---
const dbFile = "directory.db";
const extractedSchema = await getDatabaseSchema(dbFile);
console.log(extractedSchema);

File: archive/python/extract_schema.py
import sqlite3


def get_database_schema(db_path: str) -> str:
  """Connects to the SQLite database and dynamically extracts the schema for all tables."""
  conn = sqlite3.connect(db_path)
  cursor = conn.cursor()

  # Get a list of all tables in the database
  cursor.execute("SELECT name FROM sqlite_master WHERE type='table';")
  tables = cursor.fetchall()

  schema_lines = []

  for table_tuple in tables:
    table_name = table_tuple[0]
    schema_lines.append(f"Table name: {table_name}")
    schema_lines.append("Columns:")

    # Fetch column details for each table
    cursor.execute(f"PRAGMA table_info('{table_name}');")
    columns = cursor.fetchall()

    for col in columns:
      col_id, col_name, col_type, not_null, default_val, pk = col
      # Format type neatly; default to TEXT if empty
      col_type_str = col_type if col_type else "TEXT"
      schema_lines.append(f"  - {col_name} ({col_type_str})")

    schema_lines.append("")  # Blank line between tables

  conn.close()
  return "\n".join(schema_lines)


# --- Example Usage ---
db_file = "directory.db"
extracted_schema = get_database_schema(db_file)
print(extracted_schema)